### ISSUE TITLE
using a decorator on an overloaded function where the decorator uses `Concatenate` causes false positive "Overloaded function implementation does not accept all possible arguments" error

### ISSUE THREAD

**Issue by DetachHead** [python/mypy#17459]
```py
from typing import Callable, Concatenate, TypeVar, overload, ParamSpec

T = TypeVar("T")
P = ParamSpec("P")
R = TypeVar("R")


def foo(fn: Callable[Concatenate[T, P], R]) -> Callable[Concatenate[T, P], R]:
    return fn


class Foo:
    @overload
    def bar(self, a: str) -> str: ...
    @overload
    def bar(self, a: int) -> int: ...

    # error: Overloaded function implementation does not accept all possible arguments of signature 1
    # error: Overloaded function implementation does not accept all possible arguments of signature 2
    @foo
    def bar(self, a: object) -> object: ...
```
[playground](https://mypy-play.net/?mypy=latest&python=3.12&flags=show-error-codes%2Cstrict%2Ccheck-untyped-defs%2Cdisallow-any-decorated%2Cdisallow-any-expr%2Cdisallow-any-explicit%2Cdisallow-any-generics%2Cdisallow-any-unimported%2Cdisallow-incomplete-defs%2Cdisallow-subclassing-any%2Cdisallow-untyped-calls%2Cdisallow-untyped-decorators%2Cdisallow-untyped-defs%2Cno-implicit-reexport%2Clocal-partial-types%2Cwarn-redundant-casts%2Cwarn-return-any%2Cwarn-unreachable%2Cwarn-unused-configs%2Cwarn-unused-ignores&gist=f12c4be02d47e7ba9768f55e4f9b4834)

**Comment by ilevkivskyi** [comment]
This is because `foo()` destroys the argument _name_ for the first argument. Internally, I see the inferred type for implementation is
```python
def (__main__.Foo, a: builtins.object) -> builtins.object
```
so it cannot accept call like `Foo.bar(self=Foo(), a=1)`. This seems way too pedantic however.

**Comment by DetachHead** [comment]
imo `self` should be implicitly positional only. i can't think of a case where you'd want to specify it as a keyword argument

**Comment by erictraut** [comment]
> imo self should be implicitly positional only

There was a discussion about this in the [typing forum](https://discuss.python.org/t/ambiguities-about-positional-only-parameters/42328). The general consensus was aligned with this viewpoint, and this is what I've implemented in pyright, but this hasn't yet been formally incorporated into the typing spec. You're welcome to add your input to the discussion thread.
